File: OpenDentBusiness/TableTypes/HL7DefField.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;

namespace OpenDentBusiness{
	///<summary>Multiple fields per segment.</summary>
	[Serializable()]
	public class HL7DefField:TableBase{
		///<summary>Primary key.</summary>
		[CrudColumn(IsPriKey=true)]
		public long HL7DefFieldNum;
		///<summary>FK to HL7DefSegment.HL7DefSegmentNum</summary>
		public long HL7DefSegmentNum;
		///<summary>Position within the segment.</summary>
		public int OrdinalPos;
		///<summary>HL7 table Id, if applicable. Example: 0234. Example: 1234/2345.  DataType will be ID.</summary>
		public string TableId;
		///<summary>The DataTypeHL7 enum will be unlinked from the db by storing as string in db. As it's loaded into OD, it will become an enum.</summary>
		[CrudColumn(SpecialType=CrudSpecialColType.EnumAsString)]
		public DataTypeHL7 DataType;
		///<summary>User will get to pick from a list of fields that we will maintain. Example: guar.nameLFM, prov.provIdName, or pat.addressCityStateZip.  See below for the full list.  This will be blank if this is a fixed text field.</summary>
		public string FieldName;
		///<summary>User will need to insert fixed text for some fields.  Either FixedText or FieldName will have a value, not both.</summary>
		public string FixedText;

		///<summary></summary>
		public HL7DefField Clone() {
			return (HL7DefField)this.MemberwiseClone();
		}

	}

	public class FieldNameAndType{
		public string Name;
		public string TableId;
		public DataTypeHL7 DataType;

		public FieldNameAndType(string name,DataTypeHL7 datatype) {
			Name=name;
			DataType=datatype;
			TableId="";
		}

		public FieldNameAndType(string name,DataTypeHL7 datatype,string tableid) {
			Name=name;
			DataType=datatype;
			TableId=tableid;
		}

		public static List<FieldNameAndType> GetFullList() {
			List<FieldNameAndType> retVal=new List<FieldNameAndType>();
			retVal.Add(new FieldNameAndType("apt.AptNum",DataTypeHL7.CX));
			retVal.Add(new FieldNameAndType("apt.lengthStartEnd",DataTypeHL7.TQ));
			retVal.Add(new FieldNameAndType("apt.Note",DataTypeHL7.CWE));
			retVal.Add(new FieldNameAndType("dateTime.Now",DataTypeHL7.DTM));
			retVal.Add(new FieldNameAndType("eventType",DataTypeHL7.ID,"0003"));
			retVal.Add(new FieldNameAndType("guar.addressCityStateZip",DataTypeHL7.XAD));
			retVal.Add(new FieldNameAndType("guar.birthdateTime",DataTypeHL7.DTM));
			retVal.Add(new FieldNameAndType("guar.ChartNumber",DataTypeHL7.CX));
			retVal.Add(new FieldNameAndType("guar.Gender",DataTypeHL7.IS));
			retVal.Add(new FieldNameAndType("guar.HmPhone",DataTypeHL7.XTN));
			retVal.Add(new FieldNameAndType("guar.nameLFM",DataTypeHL7.XPN));
			retVal.Add(new FieldNameAndType("guar.PatNum",DataTypeHL7.CX));
			retVal.Add(new FieldNameAndType("guar.SSN",DataTypeHL7.ST));
			retVal.Add(new FieldNameAndType("guar.WkPhone",DataTypeHL7.XTN));
			retVal.Add(new FieldNameAndType("messageType",DataTypeHL7.MSG));
			retVal.Add(new FieldNameAndType("pat.addressCityStateZip",DataTypeHL7.XAD));
			retVal.Add(new FieldNameAndType("pat.birthdateTime",DataTypeHL7.DTM));
			retVal.Add(new FieldNameAndType("pat.ChartNumber",DataTypeHL7.CX));
			retVal.Add(new FieldNameAndType("pat.FeeSched",DataTypeHL7.ST));
			retVal.Add(new FieldNameAndType("pat.Gender",DataTypeHL7.IS,"0001"));//M,F,U,etc.
			retVal.Add(new FieldNameAndType("pat.HmPhone",DataTypeHL7.XTN));
			retVal.Add(new FieldNameAndType("pat.nameLFM",DataTypeHL7.XPN));
			retVal.Add(new FieldNameAndType("pat.PatNum",DataTypeHL7.CX));
			retVal.Add(new FieldNameAndType("pat.Position",DataTypeHL7.CWE,"0002"));
			retVal.Add(new FieldNameAndType("pat.Race",DataTypeHL7.CWE,"0005"));
			retVal.Add(new FieldNameAndType("pat.SSN",DataTypeHL7.ST));
			retVal.Add(new FieldNameAndType("pat.WkPhone",DataTypeHL7.XTN));
			retVal.Add(new FieldNameAndType("pdfDescription",DataTypeHL7.ST));
			retVal.Add(new FieldNameAndType("pdfDataAsBase64",DataTypeHL7.ST));
			retVal.Add(new FieldNameAndType("proc.DiagnosticCode",DataTypeHL7.CWE,"0051"));
			retVal.Add(new FieldNameAndType("proc.procDateTime",DataTypeHL7.DTM));
			retVal.Add(new FieldNameAndType("proc.ProcFee",DataTypeHL7.CP));
			retVal.Add(new FieldNameAndType("proc.toothSurfRange",DataTypeHL7.CNE,"0340"));
			retVal.Add(new FieldNameAndType("proccode.ProcCode",DataTypeHL7.CNE,"0088"));
			retVal.Add(new FieldNameAndType("prov.provIdNameLFM",DataTypeHL7.XCN));//Provider id table is user defined table and different number depending on what segment it is pulled from.  Example: FT1 Performed By Code table is 0084, 
			retVal.Add(new FieldNameAndType("prov.provIdName",DataTypeHL7.XCN));
			retVal.Add(new FieldNameAndType("separators^~\\&",DataTypeHL7.ST));
			retVal.Add(new FieldNameAndType("sequenceNum",DataTypeHL7.SI));
			return retVal;
		}

		public static DataTypeHL7 GetTypeFromName(string name) {
			List<FieldNameAndType> list=GetFullList();
			for(int i=0;i<list.Count;i++) {
				if(list[i].Name==name) {
					return list[i].DataType;
				}
			}
			throw new ApplicationException("Unknown field name: "+name);
		}

		public static string GetTableNumFromName(string name) {
			List<FieldNameAndType> list=GetFullList();
			for(int i=0;i<list.Count;i++) {
				if(list[i].Name==name) {
					return list[i].TableId;
				}
			}
			throw new ApplicationException("Unknown field name: "+name);
		}
	}

	///<summary>Data types are listed in HL7 docs section 2.15.  The items in this enumeration can be freely rearranged without damaging the database.  But can't change spelling or remove existing item.</summary>
	public enum DataTypeHL7 {
		///<summary>Coded with no exceptions.  Examples: ProcCode (Dxxxx) or TreatmentArea like tooth^surface</summary>
		CNE,
		///<summary>Composite price.  Example: 125.00</summary>
		CP,
		///<summary>Coded with exceptions.  Example: Race: American Indian or Alaska Native,Asian,Black or African American,Native Hawaiian or Other Pacific,White, Hispanic,Other Race.</summary>
		CWE,
		///<summary>Extended composite ID with check digit.  Example: patient.PatNum or patient.ChartNumber or appointment.AptNum.</summary>
		CX,
		///<summary>Date.  yyyyMMdd.  4,6,or 8</summary>
		DT,
		///<summary>Date/Time.  yyyyMMddHHmmss etc.  Allowed 4,6,8,10,12,14.  Possibly more, but unlikely.</summary>
		DTM,
		///<summary>Entity identifier.  Example: appointment.AptNum</summary>
		EI,
		/// <summary>Hierarchic designator.  Application identifier.  Example: "OD" for OpenDental.</summary>
		HD,
		/// <summary>Coded value for HL7 defined tables.  Must include TableId.  Example: 0003 is eCW's event type table id.</summary>
		ID,
		///<summary>Coded value for user-defined tables.  Example: Administrative Sex, F=Female, M=Male,U=Unknown.</summary>
		IS,
		/// <summary>Message type.  Example: composed of messageType^eventType like DFT^P03</summary>
		MSG,
		/// <summary>Numeric.  Example: transaction quantity of 1.0</summary>
		NM,
		/// <summary>Processing type.  Examples: P-Production, T-Test.</summary>
		PT,
		///<summary>Sequence ID.  Example: for repeating segments number that begins with 1.</summary>
		SI,
		///<summary>String, alphanumeric.  Example: SSN or patient.FeeSchedule</summary>
		ST,
		///<summary>Timing quantity.  Example: for eCW appointment ^^duration^startTime^endTime like ^^1200^20120101112000^20120101114000 for 20 minute (1200 second) appointment starting at 11:20 on 01/01/2012</summary>
		TQ,
		/// <summary>Version identifier.  Example: 2.3</summary>
		VID,
		///<summary>Extended address.  Example: Addr1^Addr2^City^State^Zip like 120 Main St.^Suite 3^Salem^OR^97302</summary>
		XAD,
		///<summary>Extended composite ID number and name for person.  Example: provider.EcwID^provider.LName^provider.FName^provider.MI</summary>
		XCN,
		///<summary>Extended person name.  Composite data type.  Example: LName^FName^MI).</summary>
		XPN,
		///<summary>Extended telecommunication number.  Example: 5033635432</summary>
		XTN
	}

	
}

File: OpenDentalWeb/OpenDental/src/com/opendental/odweb/client/tabletypes/HL7DefField.java
package com.opendental.odweb.client.tabletypes;

import com.google.gwt.xml.client.Document;
import com.opendental.odweb.client.remoting.Serializing;

public class HL7DefField {
		/** Primary key. */
		public int HL7DefFieldNum;
		/** FK to HL7DefSegment.HL7DefSegmentNum */
		public int HL7DefSegmentNum;
		/** Position within the segment. */
		public int OrdinalPos;
		/** HL7 table Id, if applicable. Example: 0234. Example: 1234/2345.  DataType will be ID. */
		public String TableId;
		/** The DataTypeHL7 enum will be unlinked from the db by storing as string in db. As it's loaded into OD, it will become an enum. */
		public DataTypeHL7 DataType;
		/** User will get to pick from a list of fields that we will maintain. Example: guar.nameLFM, prov.provIdName, or pat.addressCityStateZip.  See below for the full list.  This will be blank if this is a fixed text field. */
		public String FieldName;
		/** User will need to insert fixed text for some fields.  Either FixedText or FieldName will have a value, not both. */
		public String FixedText;

		/** Deep copy of object. */
		public HL7DefField deepCopy() {
			HL7DefField hl7deffield=new HL7DefField();
			hl7deffield.HL7DefFieldNum=this.HL7DefFieldNum;
			hl7deffield.HL7DefSegmentNum=this.HL7DefSegmentNum;
			hl7deffield.OrdinalPos=this.OrdinalPos;
			hl7deffield.TableId=this.TableId;
			hl7deffield.DataType=this.DataType;
			hl7deffield.FieldName=this.FieldName;
			hl7deffield.FixedText=this.FixedText;
			return hl7deffield;
		}

		/** Serialize the object into XML. */
		public String serialize() {
			StringBuilder sb=new StringBuilder();
			sb.append("<HL7DefField>");
			sb.append("<HL7DefFieldNum>").append(HL7DefFieldNum).append("</HL7DefFieldNum>");
			sb.append("<HL7DefSegmentNum>").append(HL7DefSegmentNum).append("</HL7DefSegmentNum>");
			sb.append("<OrdinalPos>").append(OrdinalPos).append("</OrdinalPos>");
			sb.append("<TableId>").append(Serializing.escapeForXml(TableId)).append("</TableId>");
			sb.append("<DataType>").append(DataType.ordinal()).append("</DataType>");
			sb.append("<FieldName>").append(Serializing.escapeForXml(FieldName)).append("</FieldName>");
			sb.append("<FixedText>").append(Serializing.escapeForXml(FixedText)).append("</FixedText>");
			sb.append("</HL7DefField>");
			return sb.toString();
		}

		/** Sets all the variables on this object based on the values in the XML document.  Variables that are not in the XML document will be null or their default values.
		 * @param doc A parsed XML document.  Must be valid XML.  Does not need to contain a node for every variable on this object.
		 * @throws Exception DeserializeFromXml is entirely encased in a try catch and will throw exceptions if anything goes wrong. */
		public void deserialize(Document doc) throws Exception {
			try {
				if(Serializing.getXmlNodeValue(doc,"HL7DefFieldNum")!=null) {
					HL7DefFieldNum=Integer.valueOf(Serializing.getXmlNodeValue(doc,"HL7DefFieldNum"));
				}
				if(Serializing.getXmlNodeValue(doc,"HL7DefSegmentNum")!=null) {
					HL7DefSegmentNum=Integer.valueOf(Serializing.getXmlNodeValue(doc,"HL7DefSegmentNum"));
				}
				if(Serializing.getXmlNodeValue(doc,"OrdinalPos")!=null) {
					OrdinalPos=Integer.valueOf(Serializing.getXmlNodeValue(doc,"OrdinalPos"));
				}
				if(Serializing.getXmlNodeValue(doc,"TableId")!=null) {
					TableId=Serializing.getXmlNodeValue(doc,"TableId");
				}
				if(Serializing.getXmlNodeValue(doc,"DataType")!=null) {
					DataType=DataTypeHL7.values()[Integer.valueOf(Serializing.getXmlNodeValue(doc,"DataType"))];
				}
				if(Serializing.getXmlNodeValue(doc,"FieldName")!=null) {
					FieldName=Serializing.getXmlNodeValue(doc,"FieldName");
				}
				if(Serializing.getXmlNodeValue(doc,"FixedText")!=null) {
					FixedText=Serializing.getXmlNodeValue(doc,"FixedText");
				}
			}
			catch(Exception e) {
				throw e;
			}
		}

		/** Data types are listed in HL7 docs section 2.15.  The items in this enumeration can be freely rearranged without damaging the database.  But can't change spelling or remove existing item. */
		public enum DataTypeHL7 {
			/** Coded with no exceptions.  Examples: ProcCode (Dxxxx) or TreatmentArea like tooth^surface */
			CNE,
			/** Composite price.  Example: 125.00 */
			CP,
			/** Coded with exceptions.  Example: Race: American Indian or Alaska Native,Asian,Black or African American,Native Hawaiian or Other Pacific,White, Hispanic,Other Race. */
			CWE,
			/** Extended composite ID with check digit.  Example: patient.PatNum or patient.ChartNumber or appointment.AptNum. */
			CX,
			/** Date.  yyyyMMdd.  4,6,or 8 */
			DT,
			/** Date/Time.  yyyyMMddHHmmss etc.  Allowed 4,6,8,10,12,14.  Possibly more, but unlikely. */
			DTM,
			/** Entity identifier.  Example: appointment.AptNum */
			EI,
			/** Hierarchic designator.  Application identifier.  Example: "OD" for OpenDental. */
			HD,
			/** Coded value for HL7 defined tables.  Must include TableId.  Example: 0003 is eCW's event type table id. */
			ID,
			/** Coded value for user-defined tables.  Example: Administrative Sex, F=Female, M=Male,U=Unknown. */
			IS,
			/** Message type.  Example: composed of messageType^eventType like DFT^P03 */
			MSG,
			/** Numeric.  Example: transaction quantity of 1.0 */
			NM,
			/** Processing type.  Examples: P-Production, T-Test. */
			PT,
			/** Sequence ID.  Example: for repeating segments number that begins with 1. */
			SI,
			/** String, alphanumeric.  Example: SSN or patient.FeeSchedule */
			ST,
			/** Timing quantity.  Example: for eCW appointment ^^duration^startTime^endTime like ^^1200^20120101112000^20120101114000 for 20 minute (1200 second) appointment starting at 11:20 on 01/01/2012 */
			TQ,
			/** Version identifier.  Example: 2.3 */
			VID,
			/** Extended address.  Example: Addr1^Addr2^City^State^Zip like 120 Main St.^Suite 3^Salem^OR^97302 */
			XAD,
			/** Extended composite ID number and name for person.  Example: provider.EcwID^provider.LName^provider.FName^provider.MI */
			XCN,
			/** Extended person name.  Composite data type.  Example: LName^FName^MI). */
			XPN,
			/** Extended telecommunication number.  Example: 5033635432 */
			XTN
		}


}
